cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/DelayConstraint.java

```diff
@@ -16,16 +16,33 @@
 
 package co.cask.cdap.internal.app.runtime.schedule.constraint;
 
-import co.cask.cdap.internal.schedule.constraint.Constraint;
+import co.cask.cdap.internal.app.runtime.schedule.ProgramSchedule;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
 
 /**
  * A constraint which requires a certain amount of delay pass after the trigger is fired, before executing the program.
  */
-public class DelayConstraint extends Constraint {
+public class DelayConstraint extends AbstractCheckableConstraint {
 
   private final long millisAfterTrigger;
 
   public DelayConstraint(long millisAfterTrigger) {
+    Preconditions.checkArgument(millisAfterTrigger >= 0,
+                                "Configured delay '%s' must not be a negative value.", millisAfterTrigger);
     this.millisAfterTrigger = millisAfterTrigger;
   }
+
+  public ConstraintResult check(ProgramSchedule schedule, ConstraintContext context) {
+    return check(System.currentTimeMillis(), context.getJob().getCreationTime());
+  }
+
+  @VisibleForTesting
```

Comment:
This is exactly what I meant by making test easier above. Introducing this extra method just for testing making this class less readable and easier to make mistake in future (e.g. some extra logic added in the `check` method above may not get tested appropriately).